pyserini/search/reranker.py

```diff
@@ -0,0 +1,106 @@
+import enum
+from sklearn.linear_model import LogisticRegression
+from sklearn.svm import SVC
+from typing import List
+from ..vectorizer import TfidfVectorizer
+import uuid
+import os
+
+
+class ClassifierType(enum.Enum):
+    LR = 'lr'
+    SVM = 'svm'
+
+
+class FusionMethod(enum.Enum):
+    AVG = 'avg'
+
+
+class PseudoRelevanceClassifierReranker:
+    def __init__(self, lucene_index: str, clf_type: List[ClassifierType], r=10, n=100, alpha=1):
+        self.r = r
+        self.n = n
+        self.alpha = alpha
+        self.clf_type = clf_type
+
+        if len(clf_type) > 2:
+            raise Exception('Re-ranker takes at most two classifiers')
+
+        self.vectorizer = TfidfVectorizer(lucene_index, min_df=5)
+
+    def set_classifier(self, clf_type: ClassifierType):
+        if clf_type == ClassifierType.LR:
+            self.clf = LogisticRegression()
+        elif clf_type == ClassifierType.SVM:
+            self.clf = SVC(kernel='linear', probability=True)
+        else:
+            raise Exception("Invalid classifier type")
+
+    def get_prf_vectors(self, doc_ids: List[str]):
+        train_docs = doc_ids[:self.r] + doc_ids[-self.n:]
+        train_labels = [1] * self.r + [0] * self.n
+
+        train_vecs = self.vectorizer.get_vectors(train_docs)
+        test_vecs = self.vectorizer.get_vectors(doc_ids)
+
+        return train_vecs, train_labels, test_vecs
+
+    def rerank_with_classifier(self, doc_ids: List[str], search_scores: List[float]):
+        train_vecs, train_labels, test_vecs = self.get_prf_vectors(
+            doc_ids)
+
+        # classification
+        self.clf.fit(train_vecs, train_labels)
+        pred = self.clf.predict_proba(test_vecs)
+        rank_scores = self.normalize([p[1] for p in pred])
+
+        # interpolation
+        interpolated_scores = [a * self.alpha + b * (1-self.alpha)
+                               for a, b in zip(rank_scores, search_scores)]
+
+        return self.sort_dual_list(interpolated_scores, doc_ids)
+
+    def rerank(self, doc_ids: List[str], search_scores: List[float]):
```

Comment:
I think it'd make sense to take as input an array of hits?
that way, other data (e.g., the document text) can come along for the ride as well (and be used as part of reranking in other reranker implementations).

Reply:
Ok, I can do that. I was a little hesitant at first, since I don't know how to get the type of hits.